Added SignalProducer+Restart extension

smilypeda opened the pull request:
Adding the extension for SignalProducer class with a new `restart` method

Applying this operator to an input signal producer should result in an output signal producer that forwards all Next events from input and restarts the original one each time the input signal producer sends Completed event. Each restart should be delayed for `delayInterval`

Example project:
https://github.com/smilypeda/SignalProducerRestartExtension

smilypeda committed:
Added SignalProducer+Restart extension

smilypeda commented:
@filblue  

> What about adding a test?

Makes sense, will do

> What if original signal terminates with error?

It wasn't specified, the obvious way is to forward error event as well. Will it be enough?

filblue commented:
Error handling wasn't specified indeed, but it's unlikely that we would want resulting signal to freeze forever. Forwarding the error or completing the resulting signal would be ok.

smilypeda committed:
Updated SignalProducer+Restart extension to sent Completed event on Error event from the input signal producer

smilypeda commented:
@filblue I have added the new commit to this PR https://github.com/smilypeda/ReactiveCocoa/pull/1/commits/3220a68b6374c46fa3a62c9391d6f8d7f5d6fc71. It transforms the logic to send Completed event in case input signal sends Error

Also, I've updated test project with the tests for this extension, please review the commit:
https://github.com/smilypeda/SignalProducerRestartExtension/commit/e3601cbd6565f0543401de34f01bec2e2aa8cfc2

filwag commented:
@smilypeda Cool, much better than freezing. But that still doesn't fit perfectly with ReactiveCocoa conventions for operators. In most cases failures are forwarded as is, as soon as possible.

filwag commented on ReactiveCocoa/Swift/SignalProducer+Restart.swift:
```diff
@@ -0,0 +1,57 @@
+//
+//  SignalProducer+Restart.swift
```
That's really minor, but if we would really want to merge this into upstream RAC, this operator should be placed in `SignalProducer.swift` where it really belongs to. Same applies for the spec that's currently sits in a test project. It would fit nicely in this PR if added to `SignalProducerSpec.swift`.

smilypeda replied:
@filwag agreed, updated in commit https://github.com/smilypeda/ReactiveCocoa/pull/1/commits/25c745b925247bf22d7af274b1989c374ca79f28

filwag commented on ReactiveCocoa/Swift/SignalProducer+Restart.swift:
```diff
@@ -0,0 +1,57 @@
+                outerDisposable += innerDisposable
+                
+                signal.observeResult { result in
+                    if let value = result.value {
```
a simple switch over `result` instead of `if let - else if let _` construct would be more swifty and readable

smilypeda replied:
ovewritten by https://github.com/smilypeda/ReactiveCocoa/pull/1/commits/a1d573245123877ee7eab16ebf7d89144669647f

filwag commented on ReactiveCocoa/Swift/SignalProducer+Restart.swift:
```diff
@@ -0,0 +1,57 @@
+					}
+                }
+				
+                signal.observeCompleted {
```
As a suggestion, `signal.observeResult` and `signal.observeCompleted` can be replaced with a single `signal.observe` with a switch over all event types. Again, I think this will make it a bit more readable and declarative.

smilypeda replied:
@filwag updated, please see https://github.com/smilypeda/ReactiveCocoa/pull/1/commits/a1d573245123877ee7eab16ebf7d89144669647f

smilypeda commented:
@filwag  thanks for the review!

> Cool, much better than freezing. But that still doesn't fit perfectly with ReactiveCocoa conventions for operators. In most cases failures are forwarded as is, as soon as possible.

I see. I added completion on Error event because @filwag gave an alternative, see the comment https://github.com/smilypeda/ReactiveCocoa/pull/1#issuecomment-240143655

I think such a change is more interesting for testing. Do you want me to change it?

smilypeda committed:
SignalProducer+Restart extension - use signal.observe + switch for better readability

smilypeda committed:
Moved SignalProducer+Restart extension into SignalProducer.swift file + moved tests to SignalProducerSpec.swift

smilypeda committed:
Updated restart command tests upon text project comments

smilypeda commented:
@filwag updated the tests in commit https://github.com/smilypeda/ReactiveCocoa/pull/1/commits/e318331e8bc5e2936a5f62e3e6a982e7b7025584

added the improvements which you pointed in comments for test project commit https://github.com/smilypeda/SignalProducerRestartExtension/commit/e3601cbd6565f0543401de34f01bec2e2aa8cfc2

filwag commented:
@smilypeda that's fine, let's stick to the current version with completion.

filwag commented on ReactiveCocoa/Swift/SignalProducer.swift:
```diff
@@ -1966,6 +1966,59 @@ extension SignalProducerType {
+					}
+				}
+				
+				signal.observeCompleted {
```
completion is already handled in `signal.observe`. `recursiveRestartWithDelay` will be called twice on completion.

filwag commented on ReactiveCocoa/Swift/SignalProducer.swift:
```diff
@@ -1966,6 +1966,59 @@ extension SignalProducerType {
+						
+					case .Failed:
+						observer.sendCompleted()
+					default: break
```
As a side note, from our experience we find it safer to avoid having `default` cases in switches and to handle all cases explicitly. This will make the compiler point us to all places of interest when a new case is added to the enum.

smilypeda replied:
@filwag  Yeah :( Forgot to remove it, corrected in https://github.com/smilypeda/ReactiveCocoa/pull/1/commits/f21b2121f089e7330371dce3b543083de884498f

EDITED: that is an answer for https://github.com/smilypeda/ReactiveCocoa/pull/1/files/e318331e8bc5e2936a5f62e3e6a982e7b7025584#r75260894

smilypeda committed:
Removed double-handling of Completed event
